### PR TITLE
feat[tool]: add all imported modules to `-f annotated_ast` output

### PR DESCRIPTION
**Pull request by charles-cooper**
### Commit message

```
this commit adds all imported modules to `-f annotated_ast`
output. this includes recursively imported modules. this is a tooling
feature, so that tools don't need to invoke the compiler multiple
and inspect import statements in order to find the ASTs for imported
modules.
```

### Cute Animal Picture

Put a link to a cute animal picture inside the parenthesis-->

### CODE REVIEW

**Review by fubuloubu — APPROVED** [review]

**Review by pcaversaccio — APPROVED** [review]

File: vyper/compiler/output.py [hunk 1]
```diff
@@ -26,9 +29,31 @@ def build_ast_dict(compiler_data: CompilerData) -> dict:
...
+        else:
+            assert isinstance(info.typ, ModuleInfo)
+            ast = info.typ.module_t._module
+
+        assert isinstance(ast, vy_ast.Module)  # help mypy
+        # use resolved_path for uniqueness, since Module objects can actually
+        # come from multiple InputBundles (particularly builtin interfaces),
+        # so source_id is not guaranteed to be unique.
+        if ast.resolved_path in unique_modules:
+            # sanity check -- object equality
```
Comment (cyberthirst) [inline vyper/compiler/output.py:49]: `is` operator checks for identity, not equality

File: vyper/compiler/output.py [hunk 2]
```diff
@@ -26,9 +29,31 @@ def build_ast_dict(compiler_data: CompilerData) -> dict:
...
+            assert unique_modules[ast.resolved_path] is ast
+        unique_modules[ast.resolved_path] = ast
+
     annotated_ast_dict = {
         "contract_name": str(compiler_data.contract_path),
         "ast": ast_to_dict(compiler_data.annotated_vyper_module),
+        "imports": [ast_to_dict(ast) for ast in unique_modules.values()],
```
Comment (cyberthirst) [inline vyper/compiler/output.py:56]: it might be confusing to have all the imports flattened into main as this also includes indirect imports

**Comment by cyberthirst** [comment]
can we add a test for recursive imports?

**Review by cyberthirst — APPROVED** [review]

File: vyper/compiler/output.py (hunk 1 above)
Comment (charles-cooper) [reply vyper/compiler/output.py:49]: yea, that's what I meant by object ("pointer") equality

File: vyper/compiler/output.py (hunk 2 above)
Comment (charles-cooper) [reply vyper/compiler/output.py:56]: It's a way of normalization, it avoids duplicating trees in the output

**Comment by charles-cooper** [comment]
> can we add a test for recursive imports?

would you mind helping contribute these?

**Comment by cyberthirst** [comment]
why don't we add the imports also to `-f ast`?

**Comment by charles-cooper** [comment]
> with #4229 we could add the imports also to `-f ast`, right?
> 
> edit: fixed link

yes, we could

**Pull request merged**